annotated.go

```diff
@@ -89,3 +90,263 @@ func (a Annotated) String() string {
+			continue
+		}
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: fType.Out(i),
+		}
+		if i < len(rt.tags) {
+			structField.Tag = reflect.StructTag(rt.tags[i])
+		}
+		annotatedResult = append(annotatedResult, structField)
+	}
+
+	ann.Outs = []reflect.Type{reflect.StructOf(annotatedResult)}
+	return nil
+}
+
+// ResultTags is an Annotation that annotates the result(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional result.
+func ResultTags(tags ...string) Annotation {
+	return resultTagsAnnotation{tags}
+}
+
+// Annotate lets you annotate a function's paramter and returns with tags
+// without you having to declare separate struct definitions for them.
+//
+// For example,
+//   func NewGateway(ro, rw *db.Conn) *Gateway { ... }
+//   fx.Provide(
+//     fx.Annotate(
+//       NewGateway,
+//       fx.ParamTags(`name:"ro" optional:"true"`, `name:"rw"`),
+//       fx.ResultTags(`name:"foo"`),
+//     ),
+//   )
+//
+// Is equivalent to,
+//
+//  type params struct {
+//    fx.In
+//
+//    RO *db.Conn `name:"ro" optional:"true"`
+//    RW *db.Conn `name:"rw"`
+//  }
+//
+//  type result struct {
+//    fx.Out
+//
+//    GW *Gateway `name:"foo"`
+//   }
+//
+//   fx.Provide(func(p params) result {
+//     return result{GW: NewGateway(p.RO, p.RW)}
+//   })
+//
+// If a single annotation appears multiple times, all annotations
+// will be ignored.
+//
+// For example,
+//
+//  fx.Provide(
+//    fx.Annotate(
+//      NewGateWay,
+//      fx.ParamTags(`name:"ro" optional:"true"`),
+//      fx.ParamTags(`name:"rw"),
+//      fx.ResultTags(`name:"foo"`)
+//    )
+//  )
+//
+// is considered an invalid usage and will not apply any of the
+// Annotations to NewGateway.
+//
+// If more tags are given than the number of parameters/results, only
+// the ones up to the number of parameters/results will be applied.
+func Annotate(f interface{}, anns ...Annotation) interface{} {
+	fVal := reflect.ValueOf(f)
+	fType := fVal.Type()
+	numIn := fType.NumIn()
+	numOut := fType.NumOut()
+	annotations := newAnnotations(fType)
+
+	for _, ann := range anns {
+		if e := ann.apply(&annotations); e != nil {
+			return annotationError{
+				target: f,
+				err:    e,
+			}
+		}
+	}
+
+	ins := annotations.Ins
+	outs := annotations.Outs
+	newF := func(args []reflect.Value) []reflect.Value {
+		var fParams, fResults []reflect.Value
+		if annotations.annotatedIn {
+			fParams = make([]reflect.Value, numIn)
+			params := args[0]
+			for i := 0; i < numIn; i++ {
+				fParams[i] = params.Field(i + 1)
+			}
+		} else {
+			fParams = args
+		}
+		fResults = fVal.Call(fParams)
+		if annotations.annotatedOut {
+			// wrap the result in an annotated struct
+			results := reflect.New(outs[0]).Elem()
+			for i := 0; i < numOut; i++ {
+				if fResults[i].Type() == _typeOfError {
+					continue
+				}
+				results.FieldByName(fmt.Sprintf("Field%d",
+					i)).Set(fResults[i])
```

Comment:
This seems a bit iffy in that we're using a made-up field name to resolve this. Can we do fields by offset, possibly (with [Value.Field](https://pkg.go.dev/reflect#Value.Field))?

Since we skip errors, one way to do that would be to track a list of indexes `fields []int` of indexes in the generated `fx.Out` struct corresponding to the result value. That is, for each `fResults[i]`, `fields[i]` is the index of the field in `results` that we want to set.

    results.Field(fields[i]).Set(fResults[i])

Reply:
This is still valid. I think if we can keep track on `annotations` of the `results` field index mapping and use `Field(int i)`, this would feel slightly more secure. That said, we can do it in a follow up because this should still work even if it feels a bit unsafe

Reply:
thanks, I will address this in a follow-up PR.